### PR TITLE
JwtAuthn: support completing padding on forward jwt payload header

### PR DESCRIPTION
**Pull request by TAOXUY**
Signed-off-by: Xuyang Tao <taoxuy@google.com>

Commit Message: revert back https://github.com/envoyproxy/envoy/pull/16576 with adding option in filter API to enable the feature. Fix https://github.com/envoyproxy/envoy/issues/16528
Risk Level: Low
Testing: Yes
Docs Changes: Yes
Release Notes: Yes

### CODE REVIEW

**Comment by TAOXUY** [comment]
@qiwzhang

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 1]
```diff
@@ -211,6 +211,12 @@ message JwtProvider {
   // Specify the clock skew in seconds when verifying JWT time constraint,
   // such as `exp`, and `nbf`. If not specified, default is 60 seconds.
   uint32 clock_skew_seconds = 10;
+
+  // This field is effective only when `forward_payload_header` is specified.
+  // If true, complete the padding in the forwarded JWT payload binary if needed.
+  // By default, it is false and the JWT payload binary will be directly forwarded
+  // without adding padding.
+  bool complete_padding_in_forward_payload = 11;
```
Comment (qiwzhang) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:219]: how about `add_padding_in_forward_payload_header`

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 2]
```diff
@@ -211,6 +211,12 @@ message JwtProvider {
   // Specify the clock skew in seconds when verifying JWT time constraint,
   // such as `exp`, and `nbf`. If not specified, default is 60 seconds.
   uint32 clock_skew_seconds = 10;
+
+  // This field is effective only when `forward_payload_header` is specified.
+  // If true, complete the padding in the forwarded JWT payload binary if needed.
+  // By default, it is false and the JWT payload binary will be directly forwarded
+  // without adding padding.
```
Comment (qiwzhang) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:218]: The comment as:
// This field only applies to the field :ref:`forward_payload_header<envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider....>.
// When it is specified,  the base64 encoded payload will be added to the header.  Normally JWT based64 encode doesn't add padding. If this field is true,  the padding will be added to the header.

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -81,6 +81,7 @@ New Features
...
 * http: added upstream and downstream alpha HTTP/3 support! See :ref:`quic_options <envoy_v3_api_field_config.listener.v3.UdpListenerConfig.quic_options>` for downstream and the new http3_protocol_options in :ref:`http_protocol_options <envoy_v3_api_msg_extensions.upstreams.http.v3.HttpProtocolOptions>` for upstream HTTP/3.
 * jwt_authn: added support to fetch remote jwks asynchronously specified by :ref:`async_fetch <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.RemoteJwks.async_fetch>`.
+* jwt_authn: added support to enable standard padding in the forwarded JWT payload specified by :ref:`async_fetch <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.complete_padding_in_forward_payload>`.
```
Comment (qiwzhang) [inline docs/root/version_history/current.rst:84]: s/async_fetch/correct_field_name

Comment (qiwzhang) [inline docs/root/version_history/current.rst:84]: s/enable standard padding/add padding

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 1 above)
Comment (TAOXUY) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 2 above)
Comment (TAOXUY) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done

File: docs/root/version_history/current.rst (hunk 3 above)
Comment (TAOXUY) [reply docs/root/version_history/current.rst]: Done

Comment (TAOXUY) [reply docs/root/version_history/current.rst]: Done

**Review by qiwzhang — APPROVED** [review]

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 4]
```diff
@@ -211,6 +211,13 @@ message JwtProvider {
   // Specify the clock skew in seconds when verifying JWT time constraint,
   // such as `exp`, and `nbf`. If not specified, default is 60 seconds.
   uint32 clock_skew_seconds = 10;
+
+  // This field only applies to :ref:`forward_payload_header <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.forward_payload_header>`.
+  //
+  // When it is specified, the base64 encoded payload will be added to the
+  // header. Normally JWT based64 encode doesn't add padding. If this field is
+  // true, the padding will be added to the header.
+  bool add_padding_in_forward_payload_header = 11;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:220]: Maybe change the name to `pad_forward_payload_header`?

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 5]
```diff
@@ -211,6 +211,13 @@ message JwtProvider {
   // Specify the clock skew in seconds when verifying JWT time constraint,
   // such as `exp`, and `nbf`. If not specified, default is 60 seconds.
   uint32 clock_skew_seconds = 10;
+
+  // This field only applies to :ref:`forward_payload_header <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.forward_payload_header>`.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:215]: Please move this field next to `forward_payload_header` field (right after it).

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 6]
```diff
@@ -211,6 +211,13 @@ message JwtProvider {
...
+  //
+  // When it is specified, the base64 encoded payload will be added to the
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:217]: I'm not certain, but I think the wording here is wrong. IIUC, if the padding field is specified but `forward_payload_header` isn't specified, then nothing will be added to the header.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Left a few API comments.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 4 above)
Comment (TAOXUY) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 5 above)
Comment (TAOXUY) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 6 above)
Comment (TAOXUY) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Just reworded. PTAL.

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 7]
```diff
@@ -190,6 +190,16 @@ message JwtProvider {
   string forward_payload_header = 8
       [(validate.rules).string = {well_known_regex: HTTP_HEADER_NAME strict: false}];
 
+  // When :ref:`forward_payload_header <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.forward_payload_header>`
+  // is specified, the base64 encoded payload will be added to the headers.
+  // Normally JWT based64 encode doesn't add padding. If this field is true,
+  // the padding will be added to the header.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:196]:
```suggestion
  // the header will be padded.
```

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 8]
```diff
@@ -190,6 +190,16 @@ message JwtProvider {
...
+  //
+  // When :ref:`forward_payload_header <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.forward_payload_header>`
+  // is unspecified, nothing will be added into the headers and this field is
+  // ineffective.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:200]:
```suggestion
  // This field is only relevant if :ref:`forward_payload_header <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.forward_payload_header>`
  // is specified.
```

**Review by adisuissa — COMMENTED** [review]
A couple of API suggestions.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by TAOXUY** [comment]
@lizan @qiwzhang Please take a look and merge this PR.

**Review by qiwzhang — APPROVED** [review]

File: source/common/common/base64.cc [hunk 9]
```diff
@@ -234,6 +234,13 @@ std::string Base64::encode(const char* input, uint64_t length, bool add_padding)
   return ret;
 }
 
+void Base64::completePadding(std::string& encoded) {
+  if (encoded.length() % 4 != 0) {
+    std::string trailing_padding(4 - encoded.length() % 4, '=');
+    encoded.append(trailing_padding);
```
Comment (lizan) [inline source/common/common/base64.cc:240]: use `absl::StrAppend`

**Review by lizan — CHANGES_REQUESTED** [review]

File: source/common/common/base64.cc (hunk 9 above)
Comment (TAOXUY) [reply source/common/common/base64.cc:240]: absl::StrAppend is more efficient for appending more than 1 string. It should make no difference to append only one.

Comment (TAOXUY) [reply source/common/common/base64.cc]: Done. Thanks!

**Review by lizan — APPROVED** [review]

**Comment by TAOXUY** [comment]
Hi @lizan, PTAL. I think the test is flaky.

**Pull request merged**